File: sdk/packages/python/scripts/conformance.py
#!/usr/bin/env python3
"""Python SDK fixture runner: reads fixtures as JSON lines on stdin, executes
each against the SDK with a mock transport, writes one JSON result line per
fixture to stdout. Protocol shared with the TypeScript runner."""

from __future__ import annotations

import json
import sys
from datetime import datetime
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "src"))

from whodb import WhoDB  # noqa: E402
from whodb._errors import map_graphql_errors  # noqa: E402


class MockTransport:
    """Replays a fixture's scripted transcript, asserting each request."""

    def __init__(self, transcript: list[dict]):
        self._transcript = list(transcript)
        self._index = 0

    def execute(self, operation: str, document: str, variables: dict[str, Any]) -> dict[str, Any]:
        if self._index >= len(self._transcript):
            raise AssertionError(f"unexpected call #{self._index + 1}: {operation}")
        step = self._transcript[self._index]
        self._index += 1
        expect = step.get("expectRequest") or {}
        expected_operation = expect.get("operation")
        if expected_operation and expected_operation != operation:
            raise AssertionError(f"expected operation {expected_operation}, got {operation}")
        for key, value in (expect.get("variables") or {}).items():
            _assert_matches(variables.get(key), value, f"variables.{key}")
        for key, value in (expect.get("variablesContain") or {}).items():
            _assert_matches(variables.get(key), value, f"variables.{key}")
        for key, value in (expect.get("inputContains") or {}).items():
            _assert_matches((variables.get("input") or {}).get(key), value, f"input.{key}")
        response = step["response"]
        if response.get("errors"):
            raise map_graphql_errors(response["errors"])
        return response["data"]


def _assert_matches(actual: Any, expected: Any, path: str) -> None:
    if json.dumps(actual, sort_keys=True) != json.dumps(expected, sort_keys=True):
        raise AssertionError(
            f"mismatch at {path}: expected {json.dumps(expected)}, got {json.dumps(actual)}"
        )


def _canonical(value: Any) -> Any:
    """Serialize results for comparison; datetimes become '@date:<iso Z>'."""
    if isinstance(value, datetime):
        iso = value.isoformat().replace("+00:00", "Z")
        return f"@date:{iso}"
    if isinstance(value, list):
        return [_canonical(item) for item in value]
    if isinstance(value, dict):
        return {key: _canonical(item) for key, item in value.items()}
    return value


def _snake(name: str) -> str:
    out = []
    for index, char in enumerate(name):
        if char.isupper() and index > 0:
            out.append("_")
        out.append(char.lower())
    return "".join(out)


def run_fixture(fixture: dict) -> dict:
    transport = MockTransport(fixture.get("transcript") or [])
    client = WhoDB(
        api_key="test",
        org="00000000-0000-0000-0000-000000000001",
        project="proj-1",
        transport=transport,
    )
    call = fixture["call"]
    try:
        domain = getattr(client, call["domain"])
        handle = domain(call["handle"])
        method = getattr(handle, _snake(call["method"]))
        args = call.get("args") or []
        # Fixture args are TS-shaped: trailing options objects become kwargs.
        positional = list(args)
        kwargs: dict[str, Any] = {}
        if positional and isinstance(positional[-1], dict) and call["method"] in (
            "list",
            "createMany",
            "followLink",
        ):
            options = positional.pop()
            for key, value in options.items():
                kwargs[_snake(key)] = value
        result: Any = method(*positional, **kwargs)
        if call.get("collect") == "pages":
            result = [page.rows for page in result.pages()]
        elif hasattr(result, "all") and callable(result.all):
            result = result.all()
        if fixture.get("expectError"):
            return {
                "name": fixture["name"],
                "pass": False,
                "reason": f"expected {fixture['expectError']['type']}, got success",
            }
        got = _canonical(result if result is not None else None)
        want = fixture.get("expectResult")
        if json.dumps(got, sort_keys=True) != json.dumps(want, sort_keys=True):
            return {
                "name": fixture["name"],
                "pass": False,
                "reason": f"result mismatch:\n  want {json.dumps(want)}\n  got  {json.dumps(got)}",
            }
        return {"name": fixture["name"], "pass": True}
    except Exception as error:  # noqa: BLE001 — fixture harness
        expected = fixture.get("expectError")
        if expected:
            type_ok = type(error).__name__ == expected["type"]
            message_ok = expected.get("messageContains", "") in str(error)
            if type_ok and message_ok:
                return {"name": fixture["name"], "pass": True}
            return {
                "name": fixture["name"],
                "pass": False,
                "reason": f"expected {expected['type']}({expected.get('messageContains', '')}), "
                f"got {type(error).__name__}: {error}",
            }
        return {"name": fixture["name"], "pass": False, "reason": f"{type(error).__name__}: {error}"}


def main() -> None:
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        result = run_fixture(json.loads(line))
        sys.stdout.write(json.dumps(result) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: sdk/packages/typescript/scripts/conformance.mjs
#!/usr/bin/env node
// conformance.mjs — TypeScript SDK fixture runner. Reads fixtures as JSON
// lines on stdin, executes each against the built SDK with a mock transport,
// and writes one JSON result line per fixture to stdout.
import { createInterface } from 'node:readline';
import { WhoDB } from '../dist/index.js';

/** Mock transport that replays a fixture's scripted transcript. */
class MockTransport {
  constructor(transcript) {
    this.transcript = [...transcript];
    this.callIndex = 0;
  }

  async execute(operationName, _document, variables) {
    const step = this.transcript[this.callIndex];
    if (!step) throw new Error(`unexpected call #${this.callIndex + 1}: ${operationName}`);
    this.callIndex += 1;
    const expect = step.expectRequest ?? {};
    if (expect.operation && expect.operation !== operationName) {
      throw new Error(`expected operation ${expect.operation}, got ${operationName}`);
    }
    for (const [key, value] of Object.entries(expect.variables ?? {})) {
      assertDeepContains(variables[key], value, `variables.${key}`);
    }
    for (const [key, value] of Object.entries(expect.variablesContain ?? {})) {
      assertDeepContains(variables[key], value, `variables.${key}`);
    }
    for (const [key, value] of Object.entries(expect.inputContains ?? {})) {
      assertDeepContains(variables.input?.[key], value, `input.${key}`);
    }
    const response = step.response;
    if (response.errors) {
      const { mapGraphQLErrors } = await import('../dist/errors.js');
      throw mapGraphQLErrors(response.errors);
    }
    return response.data;
  }
}

function assertDeepContains(actual, expected, path) {
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`mismatch at ${path}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

/** Serializes result values for comparison; Dates become "@date:<iso>". */
function canonical(value) {
  if (value instanceof Date) return `@date:${value.toISOString().replace('.000Z', 'Z')}`;
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value).map(([k, v]) => [k, canonical(v)]));
  }
  return value;
}

async function runFixture(fixture) {
  const transport = new MockTransport(fixture.transcript);
  const client = new WhoDB({ apiKey: 'test', org: '00000000-0000-0000-0000-000000000001', project: 'proj-1' }, transport);
  // proj-1 is not a UUID, but with a custom transport the workspace resolver
  // would try MyOrganizations/Projects lookups — fixtures use a UUID org and
  // treat "proj-1" as an ID by overriding below.
  try {
    const handle = client[fixture.call.domain](fixture.call.handle);
    const method = handle[fixture.call.method].bind(handle);
    let result = method(...fixture.call.args);
    if (fixture.call.collect === 'pages') {
      const pages = [];
      for await (const page of result.pages()) pages.push(page.rows);
      result = pages;
    } else {
      result = await result;
    }
    if (fixture.expectError) {
      return { name: fixture.name, pass: false, reason: `expected ${fixture.expectError.type}, got success` };
    }
    const got = canonical(result ?? null);
    const want = fixture.expectResult ?? null;
    if (JSON.stringify(got) !== JSON.stringify(want)) {
      return { name: fixture.name, pass: false, reason: `result mismatch:\n  want ${JSON.stringify(want)}\n  got  ${JSON.stringify(got)}` };
    }
    return { name: fixture.name, pass: true };
  } catch (error) {
    if (fixture.expectError) {
      const typeOk = error.constructor.name === fixture.expectError.type;
      const messageOk = !fixture.expectError.messageContains || String(error.message).includes(fixture.expectError.messageContains);
      if (typeOk && messageOk) return { name: fixture.name, pass: true };
      return { name: fixture.name, pass: false, reason: `expected ${fixture.expectError.type}(${fixture.expectError.messageContains ?? ''}), got ${error.constructor.name}: ${error.message}` };
    }
    return { name: fixture.name, pass: false, reason: `${error.constructor.name}: ${error.message}` };
  }
}

const lines = createInterface({ input: process.stdin });
for await (const line of lines) {
  if (!line.trim()) continue;
  const result = await runFixture(JSON.parse(line));
  process.stdout.write(JSON.stringify(result) + '\n');
}
